### PR TITLE
fix(fluids): Fix primary PETSc memory leak, others

### LINKED ISSUE #1490: fluids: Fix memory leak
When running with `-malloc_dump`, PETSc reports quite a few memory leaks. Judging from the earliest report:
```
[ 0] 4992 bytes
      [0]  DMCreate() at /home/jrwrigh/software/petsc/src/dm/interface/dm.c:63
      [0]  CreateDM() at /home/jrwrigh/software/libCEED/examples/fluids/src/setupdm.c:22
```
I'm guessing the `DM` is being duplicated somewhere, but not being freed. Or, more likely, it's being duplicated, attached to another struct, and that struct isn't running `DMDestroy` on it.

Note that I can go all the way back to `7fb7f898` (#1143) and the problem is still present. So it's something very fundamental that is wrong.

### PR DESCRIPTION
**Pull request by jrwrigh**
All the memory leaks reported in regression tests via `-malloc_dump` are now related to the boundary conditions and their handling of QFunctionContexts, which is a much more difficult problem and will be addressed later by implementing #1090.

Fixes #1490

### CODE REVIEW

File: examples/fluids/src/misc.c [hunk 1]
```diff
@@ -62,11 +62,12 @@ PetscErrorCode SetBCsFromICs(DM dm, Vec Q, Vec Q_loc) {
...
   PetscCall(DMGetNamedLocalVector(dm, "boundary mask", &boundary_mask));
-  PetscCall(DMGetGlobalVector(dm, &Q));
+  PetscCall(DMGetGlobalVector(dm, &ones));
   PetscCall(VecZeroEntries(boundary_mask));
-  PetscCall(VecSet(Q, 1.0));
-  PetscCall(DMGlobalToLocal(dm, Q, INSERT_VALUES, boundary_mask));
+  PetscCall(VecSet(ones, 1.0));
+  PetscCall(DMGlobalToLocal(dm, ones, INSERT_VALUES, boundary_mask));
   PetscCall(DMRestoreNamedLocalVector(dm, "boundary mask", &boundary_mask));
+  PetscCall(DMRestoreGlobalVector(dm, &ones));
```
Comment (jedbrown) [inline examples/fluids/src/misc.c:70]: Purely stylistic, but you could declare it closer.
```c
{
  Vec ones;
  PetscCall(DMGetGlobalVector(dm, &ones));
  PetscCall(VecZeroEntries(boundary_mask));
  PetscCall(VecSet(ones, 1.0));
  PetscCall(DMGlobalToLocal(dm, ones, INSERT_VALUES, boundary_mask));
  PetscCall(DMRestoreNamedLocalVector(dm, "boundary mask", &boundary_mask));
  PetscCall(DMRestoreGlobalVector(dm, &ones));
}
```

**Review by jedbrown — APPROVED** [review]
Thanks for tracking these down.

File: examples/fluids/src/misc.c (hunk 1 above)
Comment (jrwrigh) [reply examples/fluids/src/misc.c:70]: Yep, I thought about doing that too, but didn't execute for whatever reason.

File: examples/fluids/problems/newtonian.c [hunk 2]
```diff
@@ -363,8 +363,10 @@ PetscErrorCode NS_NEWTONIAN_IG(ProblemData *problem, DM dm, void *ctx, SimpleBC
   problem->apply_vol_rhs.qfunction_context = newtonian_ig_context;
   PetscCallCeed(ceed, CeedQFunctionContextReferenceCopy(newtonian_ig_context, &problem->apply_vol_ifunction.qfunction_context));
   PetscCallCeed(ceed, CeedQFunctionContextReferenceCopy(newtonian_ig_context, &problem->apply_vol_ijacobian.qfunction_context));
-  PetscCallCeed(ceed, CeedQFunctionContextReferenceCopy(newtonian_ig_context, &problem->apply_inflow.qfunction_context));
-  PetscCallCeed(ceed, CeedQFunctionContextReferenceCopy(newtonian_ig_context, &problem->apply_inflow_jacobian.qfunction_context));
+  if (bc->num_inflow > 0) {
```
Comment (jeremylt) [inline examples/fluids/problems/newtonian.c:366]: Its safe to Destroy NULL, so is there a way to get around needing so many > 0 checks?

Comment (jeremylt) [reply examples/fluids/problems/newtonian.c:366]: This sort of reduction in special casing logic would probably be the sort of thing that could spread all over the place and would probably be worth being its own branch

Comment (jrwrigh) [reply examples/fluids/problems/newtonian.c:366]: Yeah, I think so. I realized that the context destroying code in `navierstokes.c` repeated a few values, so I fixed that in an update. I'll double check if that absolves the need for the if statement here.

As for the other `> 0` checks, that necessary for the boundary condition setup in general. Something I want to get rid of with a re-write of the BC handling code. But that's for another day.

Comment (jeremylt) [reply examples/fluids/problems/newtonian.c:366]: fair, just wanted to put that on the radar since its bit me a lot in these projects

**Review by jeremylt — APPROVED** [review]
